### ISSUE TITLE
Resolving URLs returned by FindServersOnNetwork

### ISSUE THREAD

**Issue by martinboers** [OPCFoundation/UA-LDS#67]
This is not an issue specifically with the LDS, but I am having trouble finding a simple solution (or any solution) to this problem when using the LDS.

I am using LDS version 1.04.402.

We have an OPC UA server on an embedded device. Let's say the DNS name of that device is "mydevice". This OPC UA server has a discovery endpoint at opc.tcp://mydevice:4840

The OPC UA server registers itself with an LDS server running on a Windows PC. This works fine.

The call to FindServersOnNetwork returns this discovery URL for our OPC UA server: `opc.tcp://mydevice.local.:4840`

I understand that the `.local.` part of the URL returned by FindServersOnNetwork conforms to some sort of general standard, but I cannot figure out how to get the OPC UA client (UaExpert in this case) to resolve this URL when running on either (a) Debian 10 or (b) Windows 10. I expect there is some simple entry in the `hosts` file (or in another network config file) that would allow the URL `mydevice.local.` to be resolved to the IP address of `mydevice`, but after a lot of googling I still can't figure out where or what this setting should be.

If you are able to give any clues, it would be much appreciated.

**Comment by opcfoundation-org** [comment]
.local. is a top level domain defined by mDNS:
https://en.wikipedia.org/wiki/.local

The OS should resolve this domain automatically if you have the LDS installed on the client machine (the LDS installs the mDNS libraries needed to get the OS to resolve the domain on Windows – on Linux it is supposed to be a core OS feature).

I tested with ping on my Windows machine and adding .local. to other hosts on my LAN worked.

It is not clear why it is not working for you.
Does it work if you have the server running on a different windows host?

**Comment by martinboers** [comment]
> The OS should resolve this domain automatically if you have the LDS installed on the client machine (the LDS installs the mDNS libraries needed to get the OS to resolve the domain on Windows – on Linux it is supposed to be a core OS feature).

I tried the client (UaExpert) on both the LDS (Windows) machine, and a remote (Linux) machine. Same problem in both cases - the discovery endpoint returned by the LDS (`opc.tcp://mydevice.local.:4840`) is discarded by UaExpert.

On the LDS (windows) machine, I can see that the "OPCF Bonjour Service" is running.

I have not configured any DNS server on the local network. DNS entries for each of the three network devices - OPC UA Server, LDS and OPC UA Client - have simply been added to the `hosts` file on each device.

> I tested with ping on my Windows machine and adding .local. to other hosts on my LAN worked.

On the LDS (windows) machine I tried:

```cmd
C:\Users\iebm01>ping mydevice

Pinging mydevice [192.168.178.142] with 32 bytes of data:
Reply from 192.168.178.142: bytes=32 time=2ms TTL=64
Reply from 192.168.178.142: bytes=32 time=5ms TTL=64

C:\Users\iebm01>ping mydevice.local.
Ping request could not find host mydevice.local.. Please check the name and try again.
```

> Does it work if you have the server running on a different windows host?

I don't have another windows host to test with at the moment (lockdown etc.), but I am not sure why that would work differently to what I am seeing.

> Also make sure mDNS is actually installed on your Debian machine:
> https://wiki.debian.org/ZeroConf

Yes, I came across this in my earlier googling. I am not sure what advice to take from that page, except to install `avahi-daemon`, `avahi-discover` and `libnss-mdns`, which were already installed on my Debian machine. I guess that this fulfils the same role as Bonjour on Windows. And I get the same result:

```bash
$ ping mydevice 
PING mydevice (192.168.178.142) 56(84) bytes of data.
64 bytes from mydevice (192.168.178.142): icmp_seq=1 ttl=64 time=2.84 ms
64 bytes from mydevice (192.168.178.142): icmp_seq=2 ttl=64 time=4.32 ms

$ ping mydevice.local.
ping: mydevice.local.: Name or service not known
```

In both cases I guess that I am missing some configuration setting(s) to get `.local.` DNS names resolved.

**Comment by martinboers** [comment]
I think I have found the issue on my Debian 10 machine.

When I said `hosts` file, what I meant was `/etc/hosts`, because I assumed there was only one `hosts` file on the system - that's me knowing almost nothing about Ethernet network administration.

I have now discovered the `/etc/avahi/hosts` file, and adding a DNS entry for `mydevice.local` to this file (and restarting `avahi-daemon`) seems to have done the trick. I can now ping `mydevice`, `mydevice.local` and `mydevice.local.` from the Debian machine, so I am guessing that UaExpert will now be able to resolve the DNS name returned by the LDS.

I'm not so fussed about Windows, but I came across [this post](https://superuser.com/questions/1330027/how-to-enable-mdns-on-windows-10-build-17134/1330028) that might help explain what you are seeing, I'm not sure.

Thanks a million for taking the time to look at this question.
